src/Client.sol

```diff
@@ -0,0 +1,175 @@
+    {
+        uint256 operatorDataLength;
+        uint256 allocationRequestsLength;
+        uint64 provider;
+
+        (operatorDataLength, byteIdx) = CBORDecoder.readFixedArray(cborData, byteIdx);
+        if (operatorDataLength != 2) revert Errors.InvalidOperatorData();
+
+        (allocationRequestsLength, byteIdx) = CBORDecoder.readFixedArray(cborData, byteIdx);
+
+        providers = new uint64[](allocationRequestsLength);
+
+        for (uint256 i = 0; i < operatorDataLength; i++) {
+            uint256 allocationRequestLength;
+            (allocationRequestLength, byteIdx) = CBORDecoder.readFixedArray(cborData, byteIdx);
+
+            if (allocationRequestLength != 6) {
+                revert Errors.InvalidAllocationRequest();
+            }
+
+            (provider, byteIdx) = CBORDecoder.readUInt64(cborData, byteIdx);
+            providers[i] = provider;
+
+            (, byteIdx) = CBORDecoder.readBytes(cborData, byteIdx);
+            (, byteIdx) = CBORDecoder.readUInt64(cborData, byteIdx);
+            (, byteIdx) = CBORDecoder.readInt64(cborData, byteIdx);
+            (, byteIdx) = CBORDecoder.readInt64(cborData, byteIdx);
+            (, byteIdx) = CBORDecoder.readInt64(cborData, byteIdx);
+        }
+    }
+
+    /**
+     * @dev This function matches the providers with the allowedSPs
+     * @param providers The providers list to match
+     */
+    function _matchAllowedSPs(uint64[] memory providers) internal view {
+        for (uint256 i = 0; i < providers.length; i++) {
+            if (!clientSPs[msg.sender][providers[i]]) {
+                revert Errors.NotAllowedSP();
+            }
+        }
+    }
+
+    /**
+     * @notice Increase client allowance
+     * @param client Client that will receive allowance
+     * @param amount Amount of allowance to add
+     * @dev Emits AllowanceChanged event
+     * @dev Reverts if not called by manager
+     * @dev Reverts if trying to increase allowance by 0
+     */
+    function increaseAllowance(address client, uint256 amount) external onlyManager {
+        if (amount == 0) revert Errors.AmountEqualZero();
+        uint256 allowanceBefore = allowances[client];
+        allowances[client] += amount;
+        emit AllowanceChanged(client, allowanceBefore, allowances[client]);
+    }
+
+    /**
+     * @notice Decrease client allowance
+     * @param client Client whose allowance is reduced
+     * @param amount Amount to decrease the allowance
+     * @dev Emits AllowanceChanged event
+     * @dev Reverts if not called by manager
+     * @dev Reverts if trying to decrease allowance by 0
+     * @dev Reverts if amount is greater than the current client allowance
+     */
+    function decreaseAllowance(address client, uint256 amount) external onlyManager {
+        if (amount == 0) revert Errors.AmountEqualZero();
+        uint256 allowanceBefore = allowances[client];
+        if (allowanceBefore < amount) revert Errors.AmountToDecreaseBiggerThenClientAllowance();
```

Comment:
Actually... this can be used to front-run decreasing allowance. Scenario:
1. Client has 100 datacap
2. Manager wants to take it away, so submits a transaction decreaseAllowance(client, 100) which gets put in mempool
3. Attacker sees that, so the attacker sends a tx spending 1 datacap with higher gas price than manager's tx
4. Attacker's tx is executed first, so it changes the allowance to `99`
5. Managers transaction is reverted, because `99 < 100`
6. Attacker still has 99 allowance, even though manager tried to take it away

Solution: just reduce all the way to zero if amount > allowanceBefore.